src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,173 @@
+
+        bool operator==(const SlidingWindow& r2) const;
+        bool operator!=(const SlidingWindow& r2) const;
+        T& operator[](UInt index);
+        const T& operator[](UInt index) const;
+
+
+      private:
+        std::vector<T> buffer_;
+        UInt idxNext_;
+        size_t size_;
+}; 
+}} //end ns
+
+/// IMPLEMENTATION
+#include <cmath>
+
+using nupic::util::SlidingWindow;
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(nupic::UInt maxCapacity) :
+  maxCapacity(maxCapacity) {
+  NTA_CHECK(maxCapacity > 0);
+  buffer_.reserve(maxCapacity);
+  buffer_.resize(maxCapacity);
+  idxNext_ = 0;
+  size_ = 0;
+}
+
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(nupic::UInt maxCapacity, std::vector<T> initialData)  :
+  SlidingWindow(maxCapacity) {
+  auto sz = std::min(initialData.size(), (size_t)maxCapacity);
+  buffer_.insert(std::begin(buffer_), std::end(initialData) - sz, std::end(initialData));
+  buffer_.resize(sz);
+  idxNext_ = sz % maxCapacity;
+  size_ = sz;
+}
+
+
+template<class T>
+size_t SlidingWindow<T>::size() const {
+  NTA_ASSERT(buffer_.size() <= maxCapacity);
+  return std::min(size_, (size_t)maxCapacity);
+}
+
+
+template<class T>
+void SlidingWindow<T>::append(T newValue) {
+    buffer_[idxNext_] = newValue;
+    idxNext_ = ++idxNext_ %maxCapacity;//the assignment must be out of the [] above -not [idxNext_++%maxCap]- because we want to store the value %maxCap, not only ++
+  ++size_;
+}
+
+
+template<class T>
+T SlidingWindow<T>::append(T newValue, bool& isValid) {
+  T old = buffer_.empty() ? newValue : buffer_[idxNext_]; //handle case of empty buffer (access buff[0]), otherwise return oldest elem
+  isValid = (buffer_.size()==maxCapacity); //only in this case we drop oldest; this happens always after first maxCap steps ; must be checked before append()
+  append(newValue);
+  return old; 
+}
+
+
+template<class T>
+const std::vector<T>& SlidingWindow<T>::getData() const {
+  return buffer_; //may contain trailing "zeros"
+}
+
+
+template<class T>
+std::vector<T> SlidingWindow<T>::getLinearizedData() const {
+  std::vector<T> lin;
+  lin.reserve(buffer_.size());
+
+  lin.insert(std::begin(lin), std::begin(buffer_) + idxNext_, std::end(buffer_)); //insert the "older" part at the beginning
+  lin.insert(std::end(lin), std::begin(buffer_), std::begin(buffer_) + idxNext_); //append the "newer" part to the end of the constructed vect
+ return lin;
+}
+
+
+template<class T>
+bool SlidingWindow<T>::operator==(const SlidingWindow& r2) const //FIXME review the ==, on my machine it randomly passes/fails the test!
+{
+  return ((this->size() == r2.size()) && (this->maxCapacity == r2.maxCapacity) && 
+    (this->getData()== r2.getData()) );
+}
+
+
+template<class T>
+bool SlidingWindow<T>::operator!=(const SlidingWindow& r2) const
+{
+  return !operator==(r2);
+}
+
+
+template<class T> 
+T& SlidingWindow<T>::operator[](nupic::UInt index) {
+  NTA_ASSERT(index <= size());
+  return &buffer_[index];
```

Comment:
This `operator[]` should return the `i`th value in the linearized data, not the buffer. This gives people a way to iterate over the values in order without copying the vector. (Adding a `MovingWindow::Iterator` would provide a second way of doing this.)